src/hamiltonianfilereader.cpp

```diff
@@ -36,28 +40,58 @@ void HamiltonianFileReader::receiveHsys(Mat& Ad, Mat& Bd){
         PetscInt row, col;
         double real, imag;
         if (!(iss >> row >> col >> real >> imag)) continue;
-        // Assemble: Ad = Real(-i*Hsys) = Imag(Hsys)
-        // Assemble: Bd = Imag(-i*Hsys) = -Real(Hsys)
-        if (lindbladtype == LindbladType::NONE) {
-            // Schroedinger 
-            if (fabs(imag) > 1e-15) MatSetValue(Ad, row, col, imag, INSERT_VALUES);
-            if (fabs(real) > 1e-15) MatSetValue(Bd, row, col, -real, INSERT_VALUES);
-        } else {
-            // Lindblad: Vectorize I_N \kron X - X^T \kron I_N
-            for (PetscInt k = 0; k < dim_rho; k++) {
-              PetscInt rowk = row + dim_rho * k;
-              PetscInt colk = col + dim_rho * k;
-              if (fabs(imag) > 1e-15) MatSetValue(Ad, rowk, colk, imag, INSERT_VALUES);
-              if (fabs(real) > 1e-15) MatSetValue(Bd, rowk, colk, -real, INSERT_VALUES);
-              rowk = col * dim_rho + k;
-              colk = row * dim_rho + k;
-              if (fabs(imag) > 1e-15) MatSetValue(Ad, rowk, colk, -imag, INSERT_VALUES);
-              if (fabs(real) > 1e-15) MatSetValue(Bd, rowk, colk, real, INSERT_VALUES);
-            }
-        }
+        rows.push_back(row);
+        cols.push_back(col);
+        real_vals.push_back(real);
+        imag_vals.push_back(imag);
     }
     infile.close();
   }
+
+  // Broadcast the size and data to all ranks
+  int num_entries = rows.size();
+  MPI_Bcast(&num_entries, 1, MPI_INT, 0, MPI_COMM_WORLD);
+
+  if (mpirank_world != 0) {
+    rows.resize(num_entries);
+    cols.resize(num_entries);
+    real_vals.resize(num_entries);
+    imag_vals.resize(num_entries);
+  }
+
+  MPI_Bcast(rows.data(), num_entries, MPI_INT, 0, MPI_COMM_WORLD);
+  MPI_Bcast(cols.data(), num_entries, MPI_INT, 0, MPI_COMM_WORLD);
```

Comment:
Since `rows` and `cols` have `PetscInt`s, I think we should use `MPIU_INT`
```suggestion
  MPI_Bcast(rows.data(), num_entries, MPIU_INT, 0, MPI_COMM_WORLD);
  MPI_Bcast(cols.data(), num_entries, MPIU_INT, 0, MPI_COMM_WORLD);
```